File: src/main/java/net/skyenetwork/skyecrates/crates/CrateManager.java
package net.skyenetwork.skyecrates.crates;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.skyenetwork.skyecrates.SkyeCratesPlugin;
import org.bukkit.*;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;

import java.io.File;
import java.util.*;

public class CrateManager {
    
    private final SkyeCratesPlugin plugin;
    private final Map<String, CrateConfig> crates;
    private final Map<Location, String> placedCrates;
    private final Map<UUID, Map<String, Integer>> playerKeys; // Player UUID -> Crate Name -> Key Count
    
    public CrateManager(SkyeCratesPlugin plugin) {
        this.plugin = plugin;
        this.crates = new HashMap<>();
        this.placedCrates = new HashMap<>();
        this.playerKeys = new HashMap<>();
    }
    
    public void loadCrates() {
        crates.clear();
        
        File cratesDir = new File(plugin.getDataFolder(), "crates");
        if (!cratesDir.exists()) {
            cratesDir.mkdirs();
        }
        
        // Copy default example crate if no crates exist
        File[] crateFiles = cratesDir.listFiles((dir, name) -> name.endsWith(".yml"));
        if (crateFiles == null || crateFiles.length == 0) {
            plugin.saveResource("crates/example.yml", false);
        }
        
        crateFiles = cratesDir.listFiles((dir, name) -> name.endsWith(".yml"));
        if (crateFiles != null) {
            for (File file : crateFiles) {
                loadCrate(file);
            }
        }
        
        plugin.getLogger().info("Loaded " + crates.size() + " crate(s)");
    }
    
    private void loadCrate(File file) {
        try {
            YamlConfiguration config = YamlConfiguration.loadConfiguration(file);
            String crateName = file.getName().replace(".yml", "");
            
            String name = config.getString("name", crateName);
            
            // Load particle config
            Particle particleType;
            try {
                String particleTypeString = config.getString("particles.type", "FLAME");
                particleType = Particle.valueOf(particleTypeString.toUpperCase());
            } catch (IllegalArgumentException e) {
                plugin.getLogger().warning("Invalid particle type in " + file.getName() + ", using FLAME");
                particleType = Particle.FLAME;
            }
            
            ParticleConfig particles = new ParticleConfig(
                particleType,
                config.getDouble("particles.deltaX", 0.5),
                config.getDouble("particles.deltaY", 0.5),
                config.getDouble("particles.deltaZ", 0.5),
                config.getInt("particles.count", 15),
                config.getDouble("particles.speed", 0.1)
            );
            
            // Load loot table
            String lootJson = config.getString("loot");
            if (lootJson == null) {
                plugin.getLogger().warning("No loot table found in " + file.getName());
                return;
            }
            
            // Validate JSON
            try {
                JsonParser.parseString(lootJson);
            } catch (Exception e) {
                plugin.getLogger().warning("Invalid JSON in loot table for " + file.getName() + ": " + e.getMessage());
                return;
            }
            
            CrateConfig crateConfig = new CrateConfig(name, particles, lootJson);
            crates.put(crateName, crateConfig);
            
            plugin.getLogger().info("Loaded crate: " + crateName);
            
        } catch (Exception e) {
            plugin.getLogger().severe("Failed to load crate from " + file.getName() + ": " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    public List<ItemStack> generateLootFromCrate(String crateName) {
        CrateConfig crate = getCrate(crateName);
        if (crate == null) {
            return new ArrayList<>();
        }
        return generateLootFromJson(crate.getLootJson(), new Random());
    }
    
    private List<ItemStack> generateLootFromJson(String lootJson, Random random) {
        List<ItemStack> items = new ArrayList<>();
        
        try {
            JsonObject jsonObject = JsonParser.parseString(lootJson).getAsJsonObject();
            JsonArray pools = jsonObject.getAsJsonArray("pools");
            
            if (pools != null) {
                for (JsonElement poolElement : pools) {
                    JsonObject pool = poolElement.getAsJsonObject();
                    int rolls = pool.has("rolls") ? pool.get("rolls").getAsInt() : 1;
                    JsonArray entries = pool.getAsJsonArray("entries");
                    
                    for (int i = 0; i < rolls; i++) {
                        if (entries != null) {
                            for (JsonElement entryElement : entries) {
                                JsonObject entry = entryElement.getAsJsonObject();
                                String type = entry.has("type") ? entry.get("type").getAsString() : "";
                                
                                if ("minecraft:item".equals(type)) {
                                    String itemName = entry.has("name") ? entry.get("name").getAsString() : "";
                                    int weight = entry.has("weight") ? entry.get("weight").getAsInt() : 1;
                                    
                                    // Simple weight-based selection
                                    if (random.nextInt(100) < weight * 10) { // Adjust weight scaling as needed
                                        String materialName = itemName.replace("minecraft:", "").toUpperCase();
                                        Material material = Material.matchMaterial(materialName);
                                        if (material != null) {
                                            int amount = 1;
                                            
                                            // Handle functions (like set_count)
                                            if (entry.has("functions")) {
                                                JsonArray functions = entry.getAsJsonArray("functions");
                                                for (JsonElement functionElement : functions) {
                                                    JsonObject function = functionElement.getAsJsonObject();
                                                    String functionType = function.has("function") ? function.get("function").getAsString() : "";
                                                    
                                                    if ("minecraft:set_count".equals(functionType) && function.has("count")) {
                                                        JsonObject count = function.getAsJsonObject("count");
                                                        int min = count.has("min") ? count.get("min").getAsInt() : 1;
                                                        int max = count.has("max") ? count.get("max").getAsInt() : 1;
                                                        amount = random.nextInt(max - min + 1) + min;
                                                    }
                                                }
                                            }
                                            
                                            items.add(new ItemStack(material, amount));
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            plugin.getLogger().warning("Error parsing loot table: " + e.getMessage());
            // Fallback to a default item
            items.add(new ItemStack(Material.STONE));
        }
        
        return items;
    }
    
    public CrateConfig getCrate(String name) {
        return crates.get(name);
    }
    
    public Map<String, CrateConfig> getAllCrates() {
        return new HashMap<>(crates);
    }
    
    public void placeCrate(Location location, String crateName) {
        if (crates.containsKey(crateName)) {
            placedCrates.put(location, crateName);
            location.getBlock().setType(Material.BARRIER);
        }
    }
    
    public boolean removeCrate(Location location) {
        return placedCrates.remove(location) != null;
    }
    
    public String getCrateAt(Location location) {
        return placedCrates.get(location);
    }
    
    public Map<Location, String> getPlacedCrates() {
        return new HashMap<>(placedCrates);
    }
    
    public boolean openCrate(Location location, Player player) {
        String crateName = getCrateAt(location);
        if (crateName == null) {
            return false;
        }
        
        CrateConfig crate = getCrate(crateName);
        if (crate == null) {
            return false;
        }
        
        // Check if player has a key for this crate
        if (!hasKey(player, crateName)) {
            player.sendMessage(plugin.getConfigManager().getMessage("no-key", "crate", crate.getName()));
            return false;
        }
        
        // Use the key
        if (!useKey(player, crateName)) {
            player.sendMessage(plugin.getConfigManager().getMessage("no-key", "crate", crate.getName()));
            return false;
        }
        
        // Generate loot
        List<ItemStack> loot = generateLootFromCrate(crateName);
         // Give items to player
        for (ItemStack item : loot) {
            HashMap<Integer, ItemStack> remaining = player.getInventory().addItem(item);
            // Drop items that don't fit
            for (ItemStack dropItem : remaining.values()) {
                player.getWorld().dropItemNaturally(player.getLocation(), dropItem);
            }
        }

        // Send message (crate stays in place)
        player.sendMessage(plugin.getConfigManager().getMessage("crate-opened", "crate", crate.getName()));

        return true;
    }
    
    // Key management methods
    public void addKeys(Player player, String crateName, int amount) {
        UUID playerId = player.getUniqueId();
        playerKeys.computeIfAbsent(playerId, k -> new HashMap<>());
        playerKeys.get(playerId).merge(crateName, amount, Integer::sum);
    }
    
    public boolean hasKey(Player player, String crateName) {
        UUID playerId = player.getUniqueId();
        return playerKeys.containsKey(playerId) && 
               playerKeys.get(playerId).getOrDefault(crateName, 0) > 0;
    }
    
    public boolean useKey(Player player, String crateName) {
        UUID playerId = player.getUniqueId();
        if (!hasKey(player, crateName)) {
            return false;
        }
        
        Map<String, Integer> keys = playerKeys.get(playerId);
        int currentKeys = keys.get(crateName);
        if (currentKeys > 1) {
            keys.put(crateName, currentKeys - 1);
        } else {
            keys.remove(crateName);
        }
        
        return true;
    }
    
    public int getKeyCount(Player player, String crateName) {
        UUID playerId = player.getUniqueId();
        return playerKeys.getOrDefault(playerId, new HashMap<>()).getOrDefault(crateName, 0);
    }
    
    public Map<String, Integer> getPlayerKeys(Player player) {
        UUID playerId = player.getUniqueId();
        return new HashMap<>(playerKeys.getOrDefault(playerId, new HashMap<>()));
    }

    public static class CrateConfig {
        private final String name;
        private final ParticleConfig particles;
        private final String lootJson;
        
        public CrateConfig(String name, ParticleConfig particles, String lootJson) {
            this.name = name;
            this.particles = particles;
            this.lootJson = lootJson;
        }
        
        public String getName() {
            return name;
        }
        
        public ParticleConfig getParticles() {
            return particles;
        }
        
        public String getLootJson() {
            return lootJson;
        }
    }
    
    public static class ParticleConfig {
        private final Particle type;
        private final double deltaX;
        private final double deltaY;
        private final double deltaZ;
        private final int count;
        private final double speed;
        
        public ParticleConfig(Particle type, double deltaX, double deltaY, double deltaZ, int count, double speed) {
            this.type = type;
            this.deltaX = deltaX;
            this.deltaY = deltaY;
            this.deltaZ = deltaZ;
            this.count = count;
            this.speed = speed;
        }
        
        public Particle getType() {
            return type;
        }
        
        public double getDeltaX() {
            return deltaX;
        }
        
        public double getDeltaY() {
            return deltaY;
        }
        
        public double getDeltaZ() {
            return deltaZ;
        }
        
        public int getCount() {
            return count;
        }
        
        public double getSpeed() {
            return speed;
        }
    }
}

File: src/main/kotlin/net/skyenetwork/skyecrates/crates/CrateManager.kt
package net.skyenetwork.skyecrates.crates

import com.google.gson.JsonParser
import net.skyenetwork.skyecrates.SkyeCratesPlugin
import org.bukkit.*
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.inventory.ItemStack
import org.bukkit.loot.LootTable
import java.io.File
import java.io.StringReader

data class CrateConfig(
    val name: String,
    val particles: ParticleConfig,
    val lootTable: LootTable
)

data class ParticleConfig(
    val type: Particle,
    val deltaX: Double,
    val deltaY: Double,
    val deltaZ: Double,
    val count: Int,
    val speed: Double
)

class CrateManager(private val plugin: SkyeCratesPlugin) {
    
    private val crates = mutableMapOf<String, CrateConfig>()
    private val placedCrates = mutableMapOf<Location, String>()
    
    fun loadCrates() {
        crates.clear()
        
        val cratesDir = File(plugin.dataFolder, "crates")
        if (!cratesDir.exists()) {
            cratesDir.mkdirs()
        }
        
        // Copy default example crate if no crates exist
        if (cratesDir.listFiles()?.isEmpty() == true) {
            plugin.saveResource("crates/example.yml", false)
        }
        
        cratesDir.listFiles { file -> file.extension == "yml" }?.forEach { file ->
            loadCrate(file)
        }
        
        plugin.logger.info("Loaded ${crates.size} crate(s)")
    }
    
    private fun loadCrate(file: File) {
        try {
            val config = YamlConfiguration.loadConfiguration(file)
            val crateName = file.nameWithoutExtension
            
            val name = config.getString("name") ?: crateName
            
            // Load particle config
            val particleType = try {
                Particle.valueOf(config.getString("particles.type", "FLAME") ?: "FLAME")
            } catch (e: IllegalArgumentException) {
                plugin.logger.warning("Invalid particle type in ${file.name}, using FLAME")
                Particle.FLAME
            }
            
            val particles = ParticleConfig(
                type = particleType,
                deltaX = config.getDouble("particles.deltaX", 0.5),
                deltaY = config.getDouble("particles.deltaY", 0.5),
                deltaZ = config.getDouble("particles.deltaZ", 0.5),
                count = config.getInt("particles.count", 15),
                speed = config.getDouble("particles.speed", 0.1)
            )
            
            // Load loot table
            val lootJson = config.getString("loot")
            if (lootJson == null) {
                plugin.logger.warning("No loot table found in ${file.name}")
                return
            }
            
            // Parse and create loot table
            val lootTable = createLootTable(lootJson)
            if (lootTable == null) {
                plugin.logger.warning("Failed to create loot table for ${file.name}")
                return
            }
            
            val crateConfig = CrateConfig(name, particles, lootTable)
            crates[crateName] = crateConfig
            
            plugin.logger.info("Loaded crate: $crateName")
            
        } catch (e: Exception) {
            plugin.logger.severe("Failed to load crate from ${file.name}: ${e.message}")
            e.printStackTrace()
        }
    }
    
    private fun createLootTable(lootJson: String): LootTable? {
        return try {
            // Validate JSON first
            JsonParser.parseString(lootJson)
            
            // Create a temporary loot table file
            val tempFile = File.createTempFile("skyecrates_loot", ".json")
            tempFile.writeText(lootJson)
            
            // Use Bukkit's loot table registry
            val namespacedKey = NamespacedKey(plugin, "temp_${System.currentTimeMillis()}")
            
            // For Paper/Spigot, we need to use the server's loot table functionality
            val server = Bukkit.getServer()
            val lootTableManager = server.lootTable
            
            // Create a custom loot table that generates items from the JSON
            object : LootTable {
                override fun populateLoot(random: java.util.Random?, context: org.bukkit.loot.LootContext?): MutableCollection<ItemStack> {
                    return generateLootFromJson(lootJson, random ?: java.util.Random())
                }
                
                override fun fillInventory(inventory: org.bukkit.inventory.Inventory?, random: java.util.Random?, context: org.bukkit.loot.LootContext?) {
                    val items = populateLoot(random, context)
                    inventory?.let { inv ->
                        items.forEach { item ->
                            inv.addItem(item)
                        }
                    }
                }
                
                override fun getKey(): NamespacedKey = namespacedKey
            }
            
        } catch (e: Exception) {
            plugin.logger.severe("Failed to create loot table: ${e.message}")
            null
        }
    }
    
    private fun generateLootFromJson(lootJson: String, random: java.util.Random): MutableCollection<ItemStack> {
        val items = mutableListOf<ItemStack>()
        
        try {
            val jsonObject = JsonParser.parseString(lootJson).asJsonObject
            val pools = jsonObject.getAsJsonArray("pools")
            
            pools?.forEach { poolElement ->
                val pool = poolElement.asJsonObject
                val rolls = pool.get("rolls")?.asInt ?: 1
                val entries = pool.getAsJsonArray("entries")
                
                repeat(rolls) {
                    entries?.forEach { entryElement ->
                        val entry = entryElement.asJsonObject
                        val type = entry.get("type")?.asString
                        
                        if (type == "minecraft:item") {
                            val itemName = entry.get("name")?.asString
                            val weight = entry.get("weight")?.asInt ?: 1
                            
                            // Simple weight-based selection
                            if (random.nextInt(100) < weight * 10) { // Adjust weight scaling as needed
                                val material = Material.matchMaterial(itemName?.replace("minecraft:", "") ?: "STONE")
                                if (material != null) {
                                    var amount = 1
                                    
                                    // Handle functions (like set_count)
                                    val functions = entry.getAsJsonArray("functions")
                                    functions?.forEach { functionElement ->
                                        val function = functionElement.asJsonObject
                                        val functionType = function.get("function")?.asString
                                        
                                        if (functionType == "minecraft:set_count") {
                                            val count = function.getAsJsonObject("count")
                                            val min = count.get("min")?.asInt ?: 1
                                            val max = count.get("max")?.asInt ?: 1
                                            amount = random.nextInt(max - min + 1) + min
                                        }
                                    }
                                    
                                    items.add(ItemStack(material, amount))
                                }
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            plugin.logger.warning("Error parsing loot table: ${e.message}")
            // Fallback to a default item
            items.add(ItemStack(Material.STONE))
        }
        
        return items
    }
    
    fun getCrate(name: String): CrateConfig? = crates[name]
    
    fun getAllCrates(): Map<String, CrateConfig> = crates.toMap()
    
    fun placeCrate(location: Location, crateName: String) {
        if (crates.containsKey(crateName)) {
            placedCrates[location] = crateName
            location.block.type = Material.CHEST
        }
    }
    
    fun removeCrate(location: Location): Boolean {
        return placedCrates.remove(location) != null
    }
    
    fun getCrateAt(location: Location): String? = placedCrates[location]
    
    fun getPlacedCrates(): Map<Location, String> = placedCrates.toMap()
    
    fun openCrate(location: Location, player: org.bukkit.entity.Player): Boolean {
        val crateName = getCrateAt(location) ?: return false
        val crate = getCrate(crateName) ?: return false
        
        // Generate loot
        val loot = crate.lootTable.populateLoot(java.util.Random(), null)
        
        // Give items to player
        loot.forEach { item ->
            val remaining = player.inventory.addItem(item)
            // Drop items that don't fit
            remaining.values.forEach { dropItem ->
                player.world.dropItemNaturally(player.location, dropItem)
            }
        }
        
        // Remove crate
        removeCrate(location)
        location.block.type = Material.AIR
        
        // Send message
        player.sendMessage(plugin.configManager.getMessage("crate-opened", "crate" to crate.name))
        
        return true
    }
}
